### PR TITLE
[MPOM-289] Use properties for all plugin versions (alternate)

### PR DESCRIPTION
**Pull request by ctubbsii**
Alternate implementation to PR #155 to demonstrate my alternative
proposal with the naming convention of `version.artifactId` for
properties. The properties are sorted in the properties section.

This also includes individual properties for the plugins that use
`${maven.plugin.tools.version}` and `${surefire.version}`, so they can
be set individually, but will default to using the current properties
that set the version for multiple plugins at once.

This also fixes the M2Eclipse profile, `only-eclipse`, to remove the
upper bound on the version for maven-remote-resources-plugin, because
the version specified in this POM has already exceeded the that upper
bound, and the profile would not have been working correctly.

### CODE REVIEW

**Review by martin-g — APPROVED** [review]

File: pom.xml [hunk 1]
```diff
@@ -89,22 +89,56 @@ under the License.
     <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
     <sourceReleaseAssemblyDescriptor>source-release</sourceReleaseAssemblyDescriptor>
     <gpg.useagent>true</gpg.useagent>
-    <minimalMavenBuildVersion>3.2.5</minimalMavenBuildVersion>
+    <minimalMavenBuildVersion>3.6.3</minimalMavenBuildVersion>
```
Comment (slawekjaranowski) [inline pom.xml:92]: We have separate issue for it: https://issues.apache.org/jira/browse/MPOM-399

File: pom.xml [hunk 2]
```diff
@@ -89,22 +89,56 @@ under the License.
...
     <minimalJavaBuildVersion>1.8</minimalJavaBuildVersion>
     <maven.compiler.source>${maven.compiler.target}</maven.compiler.source>
     <maven.compiler.target>1.7</maven.compiler.target>
-    <surefire.version>3.1.0</surefire.version><!-- for surefire, failsafe and surefire-report -->
+    <surefire.version>3.1.2</surefire.version><!-- for surefire, failsafe and surefire-report -->
```
Comment (slawekjaranowski) [inline pom.xml:96]: Update should be done as separate PR and issue

File: pom.xml [hunk 3]
```diff
@@ -89,22 +89,56 @@ under the License.
...
+    <version.maven-gpg-plugin>3.1.0</version.maven-gpg-plugin>
+    <version.maven-help-plugin>3.4.0</version.maven-help-plugin>
+    <version.maven-install-plugin>3.1.1</version.maven-install-plugin>
+    <version.maven-invoker-plugin>3.5.1</version.maven-invoker-plugin>
+    <version.maven-jar-plugin>3.3.0</version.maven-jar-plugin>
+    <version.maven-javadoc-plugin>3.5.0</version.maven-javadoc-plugin>
+    <version.maven-plugin-annotations>${maven.plugin.tools.version}</version.maven-plugin-annotations>
+    <version.maven-plugin-plugin>${maven.plugin.tools.version}</version.maven-plugin-plugin>
+    <version.maven-plugin-report-plugin>${maven.plugin.tools.version}</version.maven-plugin-report-plugin>
+    <version.maven-project-info-reports-plugin>3.4.5</version.maven-project-info-reports-plugin>
```
Comment (slawekjaranowski) [inline pom.xml:122]: Next version update

**Review by slawekjaranowski — COMMENTED** [review]
Looks ok for me.

To think if we want separate PRs for other updates as I commented or link / create issues for it.

File: pom.xml (hunk 1 above)
Comment (ctubbsii) [reply pom.xml:92]: I can back that out of here.

File: pom.xml (hunk 2 above)
Comment (ctubbsii) [reply pom.xml:96]: Sure, no problem. I'll back it out.

**Review by gnodet — APPROVED** [review]

**Review by slachiewicz — APPROVED** [review]

File: pom.xml [hunk 4]
```diff
@@ -97,14 +97,48 @@ under the License.
...
+    <version.maven-project-info-reports-plugin>3.4.3</version.maven-project-info-reports-plugin>
+    <version.maven-release-plugin>3.0.1</version.maven-release-plugin>
+    <version.maven-remote-resources-plugin>3.1.0</version.maven-remote-resources-plugin>
+    <version.maven-resources-plugin>3.3.1</version.maven-resources-plugin>
+    <version.maven-scm-plugin>2.0.1</version.maven-scm-plugin>
+    <version.maven-scm-publish-plugin>3.2.1</version.maven-scm-publish-plugin>
+    <version.maven-shade-plugin>3.4.1</version.maven-shade-plugin>
+    <version.maven-site-plugin>3.12.1</version.maven-site-plugin>
+    <version.maven-source-plugin>3.3.0</version.maven-source-plugin>
+    <version.maven-surefire-plugin>${surefire.version}</version.maven-surefire-plugin>
```
Comment (olamy) [inline pom.xml:131]: as the standard is now to use `<version.artifactId>` to contain version maybe this should be done in the other way to be consistent.
```
<version.maven-surefire-plugin>xxxx</version.maven-surefire-plugin>
<surefire.version>${version.maven-surefire-plugin}</surefire.version>
```

**Review by olamy — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 4 above)
Comment (hboutemy) [reply pom.xml:131]: I see the logic, but this will break the configuration that people have done with `surefire.version` until now

then I think that the current reuse of `surefire.version` is a nicer migration path

**Review by hboutemy — APPROVED** [review]
LGTM

File: pom.xml (hunk 4 above)
Comment (olamy) [reply pom.xml:131]: Not sure to understand what this will break as the property will be still there.
Any example how this will break something?

Comment (slawekjaranowski) [reply pom.xml:131]: `surefire.version` is also used in other place, like 
```
    <version.maven-failsafe-plugin>${surefire.version}</version.maven-failsafe-plugin>
```

Comment (olamy) [reply pom.xml:131]: Frankly if you think it's ok. I have no intention to block anything. 
I just find it not consistent.

**Review by slawekjaranowski — APPROVED** [review]

File: pom.xml (hunk 4 above)
Comment (ctubbsii) [reply pom.xml:131]: @olamy,

The way it currently is, a user can consistently use the new properties, and things will work as expected.
Or, if they don't know about the new properties, but are using the old one, then it will still work as expected.

If we were to reverse these, we'd get consistency of presentation in the POM, but it would hurt the user's experience:

1. First, we'd have to figure out how to deal with the failsafe plugin's version, because `surefire.version` is a special case due to it controlling 3 plugins. (`maven.plugin.tools.version` is a similar special case). If we reverse the order, then what would we do about the other 2 affected by that same property? It's not clear.
2. Second, any existing `surefire.version` set in the user's POM would be irrelevant and ignored if the user didn't know to override it using the new property. The entire value of `surefire.version` as a property to control multiple plugins would be completely irrelevant (unless we also changed the `<plugin><version/>` value to use the other property, in which case we'd have inconsistency in the `<plugins>` section there instead of in the `<properties>` section, and we wouldn't actually solve the inconsistency you're pointing out.

The current implementation I have preserves a consistent user experience. Even if it looks a little inconsistent and out of place in the POM, there's a really good reason for it.

Comment (olamy) [reply pom.xml:131]: > @olamy,
> 
> The way it currently is, a user can consistently use the new properties, and things will work as expected. Or, if they don't know about the new properties, but are using the old one, then it will still work as expected.
> 
> If we were to reverse these, we'd get consistency of presentation in the POM, but it would hurt the user's experience:
> 
> 1. First, we'd have to figure out how to deal with the failsafe plugin's version, because `surefire.version` is a special case due to it controlling 3 plugins. (`maven.plugin.tools.version` is a similar special case). If we reverse the order, then what would we do about the other 2 affected by that same property? It's not clear.
> 2. Second, any existing `surefire.version` set in the user's POM would be irrelevant and ignored if the user didn't know to override it using the new property. The entire value of `surefire.version` as a property to control multiple plugins would be completely irrelevant (unless we also changed the `<plugin><version/>` value to use the other property, in which case we'd have inconsistency in the `<plugins>` section there instead of in the `<properties>` section, and we wouldn't actually solve the inconsistency you're pointing out.
> 
> The current implementation I have preserves a consistent user experience. Even if it looks a little inconsistent and out of place in the POM, there's a really good reason for it.

as long as a change is documented users can adjust their usage. If we follow this argument nothing can be changed for life...
we could have
`<version.maven-failsafe-plugin>${version.maven-surefire-plugin}</version.maven-failsafe-plugin>`

Anyway I already don't like this but would prefer `<maven-surefire-plugin.version>` 
But as a majority prefers something else just do it, I can live with that.
It's just a matter of preference but don't waste your time trying to convince me ;)

Comment (slawekjaranowski) [reply pom.xml:131]: I agree with @olamy we can have:

```
<version.maven-failsafe-plugin>${version.maven-surefire-plugin}</version.maven-failsafe-plugin>
...
<version.maven-plugin-report-plugin>${version.maven-plugin-plugin}</version.maven-plugin-report-plugin>
```

and so on

Comment (michael-o) [reply pom.xml:131]: `version.maven-plugin-tools` and `version.maven-surefire`, no?

Comment (slawekjaranowski) [reply pom.xml:131]: I see that we want to support backward compatibility for old properties `surefire.version` and `maven.plugin.tools.version`

yes - new property for aggregate version for many items should follow new conventions,

so we can have

```
<surefire.version>3.1.0</surefire.version> <!-- old property -->
<version.maven-surefire>${surefire.version}</version.maven-surefire> <!-- new aggregate  property -->
<version.maven-surefire-plugin>${version.maven-surefire}</version.maven-surefire-plugin>
<version.maven-failsafe-plugin>${version.maven-surefire}</version.maven-failsafe-plugin>
```

Comment (michael-o) [reply pom.xml:131]: I would not introduce new split proprties, because either Surefire entirely is updated or none. It does not make sense to mix and match. Same for Plugin Tools. It just needs to be documented.

Comment (slawekjaranowski) [reply pom.xml]: @ctubbsii, @michael-o - I updated docs and use one property for such groups

**Review by michael-o — APPROVED** [review]

**Review by slachiewicz — APPROVED** [review]

**Comment by jira-importer** [comment]
Resolve #489
